README.md
![Header](Documentation/ConvenientHeader.png)

![ConvenientImagePicker](Documentation/picture.png)
# ConvenientImagePicker
**ConvenientImagePicker** is a beautiful and simple image picker solution for iOS development written on Swift. It's a view controller that can simply present it everywhere. Excellent interaction, Mutiple selection, Photo picker, Dark mode, and so on.

[![Version](https://img.shields.io/cocoapods/v/ConvenientImagePicker.svg?style=flat)](https://cocoapods.org/pods/ConvenientImagePicker)
[![License](https://img.shields.io/cocoapods/l/ConvenientImagePicker.svg?style=flat)](https://cocoapods.org/pods/ConvenientImagePicker)
[![Platform](https://img.shields.io/cocoapods/p/ConvenientImagePicker.svg?style=flat)](https://cocoapods.org/pods/ConvenientImagePicker)
![Xcode 9.0+](https://img.shields.io/badge/Xcode-9.0%2B-blue.svg)
![iOS 8.0+](https://img.shields.io/badge/iOS-8.0%2B-blue.svg)
![Swift 4.0+](https://img.shields.io/badge/Swift-4.0%2B-orange.svg)
[![中文 README](https://img.shields.io/badge/%E4%B8%AD%E6%96%87-README-blue.svg?style=flat)](https://github.com/CLOXnu/ConvenientImagePicker/blob/master/README.zh-cn.md)

## Release Notes

the newest version is 0.2.2, Add `assetsSortKey` and `assetsSortAscending` property, the order of pictures is now customizable! thanks for @Édouard. More information see [Release Notes](ReleaseNotes.md).

## Overview

**ConvenientImagePicker** provides smooth interaction, has excellent user experience, it can display system photo album and can also display the specified images.

It is worth emphasizing that the **ConvenientImagePicker** view has precise gesture control.

![Overview](Documentation/overview.GIF)![Detailview](Documentation/detailview.GIF)

## Requirements

- iOS 9.3+
- Xcode 9.0+
- Swift 4.0+

## Installation

**ConvenientImagePicker** can be installed through [CocoaPods](http://cocoapods.org), add the following entry to your Podfile:

```ruby
pod 'ConvenientImagePicker'
```

Then run ```pod install```,

and include the image picker wherever you need it with

```swift
import ConvenientImagePicker
```

, it's really a simple way.

## Usage

When you prepare to present this image picker, we assume that you will call a function like this:

```swift
func PresentPhotoPicker()
```

Well, the most simplest version is add the following code in this function:

```swift
let pickerViewController = PickerViewController()
pickerViewController.delegate = self
pickerViewController.isSupportLandscape = true // A Boolean value that determines whether the ability of landscape exists.
self.present(pickerViewController, animated: true, completion: nil)
```

Then, you are supposed to implement ```ConvenientImagePickerDelegate``` in your own view controller:

And implement these delegate function:

```swift
func imagePickerDidCancel(_ selectedImages: [Int : UIImage])
func imageDidSelect(_ imagePicker: PickerViewController, index: Int, image: UIImage?)
func imageDidDeselect(_ imagePicker: PickerViewController, index: Int, image: UIImage?)
func imageSelectMax(_ imagePicker: PickerViewController, wantToSelectIndex: Int, wantToSelectImage: UIImage?)
```

```imagePickerDidCancel``` will inform you that user has cancelled the image picker, and return the images user has selected.

```imageDidSelect``` will inform you that user has selected an image.

```imageDidDeselect``` will inform you that user has deselected an image.

```imageSelectMax``` will inform you that user want to select an image, but this time he has selected the limit number of images.

---

You can use ```imagePicker.selectedImageCount``` in last 3 functions to get the number of the images user has selected.

Do not initialize ```pickerViewController``` outside of function ```PresentPhotoPicker```.

So far, this is the simplest usage of this pod.

## Optional Configuration

Sure, You can use more features of the image picker, or even customize it, instead of just using the default configuration.

Start with ```let pickerViewController = PickerViewController()``` :

```swift
pickerViewController.maxNumberOfSelectedImage = 50 // The maximum number of pictures allowed.
pickerViewController.allowMultipleSelection = true // A Boolean value that determines whether the picker view can mutiple selection.
pickerViewController.numberOfPictureInRow = 4 // The number of pictures in a row.
pickerViewController.intervalOfPictures = 5.0 // The interval between pictures.
pickerViewController.isSimpleMode = true // A Boolean value that determines whether the title label, count view, and close button exist.
pickerViewController.images = nil // The displayed images, it's will be photo library if nil.
pickerViewController.isDarkMode = false // A Boolean value that determines whether darkmode enable.
pickerViewController.isSwitchDarkAutomately = true // A Boolean value that determines whether darkmode can switched automately. (only iOS 13 valid)
pickerViewController.initialSelectedIndex = [0,1,2,3,4] // A set of index of selected image when the picker appears.
pickerViewController.isAnimated = true // A Boolean value that determines whether the appear animation exists.
pickerViewController.customSelectedImage // A selectedImage type value that relates to the image of selected picture.
pickerViewController.isSupportLandscape = true // A Boolean value that determines whether the ability of landscape exists.
pickerViewController.assetsSortKey = "modificationDate" // A String value that determines whether the order key of all assets.
pickerViewController.assetsSortAscending = false // A Boolean value that determines whether the order way of all assets.
```

### when 'isSimpleMode = false'

<img src="https://github.com/CLOXnu/ConvenientImagePicker/blob/master/Documentation/simpleMode=false.png" alt="simpleMode=false" align="right" width="200"/>

When ```pickerViewController.isSimpleMode = false``` appear in your configure list, you are supposed to learn about ```titleView```, ```titleLabel```, ```countLabel```, ```doneButton```, and ```titleViewEffectView```.(As shown on the right)

You can customize ```titleView```, ```titleLabel```, ```countLabel```, and ```doneButton``` when ```isSimpleMode = false```.

You can also customize ```titleViewEffectView```, ```mainView```, and ```collectionView``` regardless of the value of ```isSimpleMode```, because they are always exist.

By the way, ```decorationBar``` can be customized in the case of ```isSimpleMode = true```.

If more in-controller-override needed, ```extension PickerViewController``` is necessary.

## ⚠️Notice

* Do not forget to add ```NSPhotoLibraryUsageDescription``` in your Info.plist if you want to present a photo picker.
* Check read album permissions before using photo picker.
* Please initialize new variable ```pickerViewController``` whenever preparing to present the image picker.

## Instance

In [TextCard](https://apps.apple.com/cn/app/%E6%96%87%E5%AD%97%E5%8D%A1%E7%89%87/id1473078157), which is an iOS App has imported **ConvenientImagePicker**.

<img src="Documentation/TextBoxIconNew.png" alt="TextBoxIconNew" width="100"/>

<img src="Documentation/instance.jpg" alt="instance" width="300"/>


## License

**ConvenientImagePicker** is released under the MIT license. See [LICENSE](LICENSE.md) for details.

Thank you for your support!🙏




ReleaseNotes.md
# 0.2.2 Release Notes

2020-01-28 Update

* Add `assetsSortKey` and `assetsSortAscending` property, the order of pictures is now customizable! thanks for @Édouard.

# 0.2.1 Release Notes

2019-12-30 Update

* Fixed album permission issues.

# 0.2 Release Notes

2019-12-28 Update

* Fixed again a possible crash of some devices when camera access is not allowed.
* Change the sort method of photo collections from `creationDate` to `modificationDate`.
* Add new property `isAnimated`, a Boolean value that determines whether the appear animation is exist.
* Enable the functionality of `isAnimated` parameter in present function in your view.
* Add new property `customSelectedImage`, a selectedImage type value that relates to the image of selected picture.
* Add new property `isSupportLandscape`, a Boolean value that determines whether the ability of landscape.

# 0.1.2 Release Notes

2019-12-06 Update

* Fixed a possible crash of some devices when camera access is not allowed.
* Correct the spell error of function "imageSelectMax": ` func imageSelectMax(_:, wantToSelectIndex:, wantToSelectImage:) `.
* Change the size of thumbnail image.
* Add new property `initialSelectedIndex`, a set of index of selected image when the picker appears.

# 0.1.1 Release Notes

the second version of ConvenientImagePicker, the update entries as follows:

* Refresh image collection more fast
* Reduce the memory footprint


# 0.1.0 Release Notes

the first version of ConvenientImagePicker.


Example/Tests/Tests.swift
// https://github.com/Quick/Quick

import Quick
import Nimble
import ConvenientImagePicker

class TableOfContentsSpec: QuickSpec {
    override func spec() {
        describe("these will fail") {

            it("can do maths") {
                expect(1) == 2
            }

            it("can read") {
                expect("number") == "string"
            }

            it("will eventually fail") {
                expect("time").toEventually( equal("done") )
            }
            
            context("these will pass") {

                it("can do maths") {
                    expect(23) == 23
                }

                it("can read") {
                    expect("🐮") == "🐮"
                }

                it("will eventually pass") {
                    var time = "passing"

                    DispatchQueue.main.async {
                        time = "done"
                    }

                    waitUntil { done in
                        Thread.sleep(forTimeInterval: 0.5)
                        expect(time) == "done"

                        done()
                    }
                }
            }
        }
    }
}


PresentTransition copy.swift
//
//  PresentTransition.swift
//  ConvenientImagePicker
//
//  Created by Land on 2019/10/17.
//

import UIKit

class PresentTransition2: UIPercentDrivenInteractiveTransition, UIViewControllerAnimatedTransitioning, UIViewControllerTransitioningDelegate {
    
    var isPresent = true
    var isInteractive = false
    
    private var panGesture : UIPanGestureRecognizer!
    var PickerVC : UIViewController!
    
    var velocity: CGFloat = 0
    
    func transitionDuration(using transitionContext: UIViewControllerContextTransitioning?) -> TimeInterval {
        0.5
    }
    
    func animateTransition(using transitionContext: UIViewControllerContextTransitioning) {
        
        let container = transitionContext.containerView
//        print("in")
//        guard let fromView = transitionContext.view(forKey: .from) else {return}
//        guard let toView = transitionContext.view(forKey: .to) else {return}
//        print("in2")
//
//        let theUpper = UIView(frame: CGRect(x: 0, y: 0, width: UIScreen.main.bounds.width, height: UIScreen.main.bounds.height))
        
        if self.isPresent
        {
            //guard let fromView = transitionContext.view(forKey: .from) else {return}
            guard let toView = transitionContext.view(forKey: .to) else {return}
            guard let toVC = transitionContext.viewController(forKey: .to) else {return}
            guard let fromVC = toVC.presentingViewController else {return}
            guard let fromView = fromVC.view else {return}
            let theUpper = UIView(frame: CGRect(x: 0, y: 0, width: UIScreen.main.bounds.width, height: UIScreen.main.bounds.height))
            self.PickerVC = toVC
            
            self.panGesture = UIPanGestureRecognizer()
            self.panGesture.addTarget(self, action: #selector(Panned(_:)))
            toView.addGestureRecognizer(self.panGesture)
            
            toView.frame = CGRect(x: 0, y: UIScreen.main.bounds.height, width: UIScreen.main.bounds.width, height: 0 + 50)
            toView.layer.cornerRadius = 20
            theUpper.backgroundColor = UIColor.black
            theUpper.alpha = 0

            fromView.addSubview(theUpper)
            container.addSubview(fromView)
            container.addSubview(toView)
            
            let duration = self.transitionDuration(using: transitionContext)
            UIView.animate(withDuration: duration, delay: 0, usingSpringWithDamping: 0.8, initialSpringVelocity: velocity, options: [.allowUserInteraction], animations: {
                toView.frame = CGRect(x: 0, y: UIScreen.main.bounds.height / 2.0, width: UIScreen.main.bounds.width, height: UIScreen.main.bounds.height / 2.0 + 50)
                theUpper.alpha = 0.3
            }) { (true) in
                transitionContext.completeTransition(true)
            }
        }
        else
        {
            guard let fromView = transitionContext.view(forKey: .from) else {return}
            guard let fromVC = transitionContext.viewController(forKey: .from) else {return}
            guard let toVC = fromVC.presentedViewController else {return}
            guard let toView = toVC.view else {return}
            let theUpper = UIView(frame: CGRect(x: 0, y: 0, width: UIScreen.main.bounds.width, height: UIScreen.main.bounds.height))
            
            fromView.layer.cornerRadius = 20
            theUpper.backgroundColor = UIColor.black
            theUpper.alpha = 0.3

            toView.addSubview(theUpper)
            container.addSubview(fromView)
            container.addSubview(toView)
            
            let duration = self.transitionDuration(using: transitionContext)
            UIView.animate(withDuration: duration, delay: 0, usingSpringWithDamping: 1, initialSpringVelocity: velocity, options: [.allowUserInteraction], animations: {
                fromView.frame = CGRect(x: 0, y: UIScreen.main.bounds.height, width: UIScreen.main.bounds.width, height: 0 + 50)
                theUpper.alpha = 0
            }) { (true) in
                transitionContext.completeTransition(true)
            }
        }
    }
    
    
    @objc func Panned(_ gesture: UIPanGestureRecognizer)
    {
        if gesture.state == .began
        {
            self.isInteractive = true
            self.PickerVC.dismiss(animated: true, completion: nil)
        }
        self.update(0.5)
    }
    
    
    
    
    func animationController(forPresented presented: UIViewController, presenting: UIViewController, source: UIViewController) -> UIViewControllerAnimatedTransitioning? {
        self.isPresent = true
        return self
    }
    
    func animationController(forDismissed dismissed: UIViewController) -> UIViewControllerAnimatedTransitioning? {
        self.isPresent = false
        return self
    }
    
    func interactionControllerForPresentation(using animator: UIViewControllerAnimatedTransitioning) -> UIViewControllerInteractiveTransitioning? {
        return self.isInteractive ? self : nil
    }
    
    func interactionControllerForDismissal(using animator: UIViewControllerAnimatedTransitioning) -> UIViewControllerInteractiveTransitioning? {
        return self.isInteractive ? self : nil
    }
    
    
}


ConvenientImagePicker/Classes/PickerViewController.swift
//
//  PickerViewController.swift
//  ConvenientImagePicker
//
//  Created by Land on 2019/10/16.
//

import UIKit
import Photos

let screenHeight = { () -> CGFloat in return UIScreen.main.bounds.height }
let screenWidth = { () -> CGFloat in UIScreen.main.bounds.width }
let statusBarHeight = UIApplication.shared.statusBarFrame.height


@objc public protocol ConvenientImagePickerDelegate {
    
    func imagePickerDidCancel(_ selectedImages: [Int : UIImage])
    func imageDidSelect(_ imagePicker: PickerViewController, index: Int, image: UIImage?)
    func imageDidDeselect(_ imagePicker: PickerViewController, index: Int, image: UIImage?)
    func imageSelectMax(_ imagePicker: PickerViewController, wantToSelectIndex: Int, wantToSelectImage: UIImage?)
}

public class PickerViewController: UIViewController {
    
    public weak var delegate: ConvenientImagePickerDelegate?
    
    internal var _selectedImageCount = 0
    public var selectedImageCount: Int { get{ return _selectedImageCount} }
    
    /// The maximum number of pictures allowed.
    public var maxNumberOfSelectedImage = 50
    /// A Boolean value that determines whether the picker view can mutiple selection.
    public var allowMultipleSelection = true
    public var maxSizeOfPicture = 1000
    /// The number of pictures in a row.
    public var numberOfPictureInRow = 4
    /// The interval between pictures.
    public var intervalOfPictures: CGFloat = 5
    /// A Boolean value that determines whether the title label, count view, and close button are exist.
    public var isSimpleMode = true
    /// The displayed images, it's will be photo library if nil.
    public var images: [UIImage]?
    /// A Boolean value that determines whether darkmode enable.
    public var isDarkMode = false
    /// A Boolean value that determines whether darkmode can switched automately. (only iOS 13 valid)
    public var isSwitchDarkAutomately = true
    /// A set of index of selected image when the picker appears.
    public var initialSelectedIndex = Set<Int>()
    /// A Boolean value that determines whether the appear animation is exist.
    public var isAnimated = true
    /// A selectedImage type value that relates to the image of selected picture.
    public var customSelectedImage = selectedImage()
    /// A Boolean value that determines whether the ability of landscape.
    public var isSupportLandscape = false
    /// A String value that determines whether the order key of all assets.
    public var assetsSortKey = "modificationDate"
    /// A Boolean value that determines whether the order way of all assets.
    public var assetsSortAscending = false

//    public var backgroundColor: UIColor = UIColor.white
//    public var titleColor: UIColor = UIColor.black
//    public var titleText: String = "Please select picture"
//    public var buttonTintColor: UIColor?
//    public var buttonTitle: String = "Done"
//    public var countTextColor: UIColor = UIColor.white
    
    public class selectedImage {
        public var image: UIImage?
        public var width: CGFloat = 20
        public var height: CGFloat = 20
        public var trailing: CGFloat = 10
        public var bottom: CGFloat = 10
        
        public init(image: UIImage?, width: CGFloat?, height: CGFloat?, trailing: CGFloat?, bottom: CGFloat?) {
            self.image = image
            self.width = width ?? self.width
            self.height = height ?? self.height
            self.trailing = trailing ?? self.trailing
            self.bottom = bottom ?? self.bottom
        }
        
        public convenience init() {
            self.init(image: nil, width: nil, height: nil, trailing: nil, bottom: nil)
        }
    }
    
    
    let transition = PresentTransition()
    var isDesideToCancel = false
    
    var backView = UIView()
    public var mainView = UIView()
    public var titleView = UIView()
    public var decorationBar = UIView()
    public var countLabel = UILabel()
    public var titleLabel = UILabel()
    public var doneButton = UIButton(type: .system)
    
    public var titleViewEffectView = UIVisualEffectView()
    
    func resetViewsFrame()
    {
        self.backView.frame = CGRect(x: 0, y: 0, width: screenWidth(), height: screenHeight())
        self.mainView.frame = CGRect(x: 0, y: screenHeight(), width: screenWidth(), height: screenHeight() + 50.0)
        self.titleView.frame = CGRect(x: 0, y: 0, width: screenWidth(), height: 15)
        self.decorationBar.frame = CGRect(x: screenWidth() / 2.0 - 40, y: 5, width: 80, height: 5)
        self.countLabel.frame = CGRect(x: 15, y: 15, width: 30, height: 30)
        self.titleLabel.frame = CGRect(x: 60, y: 0, width: screenWidth() - 150, height: 60)
        self.doneButton.titleLabel?.font = UIFont.boldSystemFont(ofSize: 17)
        self.titleViewEffectView.frame = CGRect(x: 0, y: 0, width: screenWidth(), height: 60)
    }
    func resetAllView()
    {
        self.backView.frame = CGRect(x: 0, y: 0, width: screenWidth(), height: screenHeight())
        self.backView.alpha = 0
        self.backView.backgroundColor = UIColor.black
        
        self.mainView.frame = CGRect(x: 0, y: screenHeight(), width: screenWidth(), height: screenHeight() + 50.0)
        self.mainView.layer.cornerRadius = 10.0
        self.mainView.clipsToBounds = true
        self.mainView.backgroundColor = UIColor.white
        
        self.titleView.frame = CGRect(x: 0, y: 0, width: screenWidth(), height: 15)
        self.titleView.clipsToBounds = true
        self.titleView.backgroundColor = UIColor.white
        
        self.decorationBar.frame = CGRect(x: screenWidth() / 2.0 - 40, y: 5, width: 80, height: 5)
        self.decorationBar.clipsToBounds = true
        self.decorationBar.layer.cornerRadius = 2.5
        self.decorationBar.backgroundColor = UIColor.black
        
        self.countLabel.frame = CGRect(x: 15, y: 15, width: 30, height: 30)
        self.countLabel.backgroundColor = UIColor.lightGray
        self.countLabel.textColor = UIColor.white
        self.countLabel.textAlignment = .center
        self.countLabel.font = UIFont.boldSystemFont(ofSize: 20)
        self.countLabel.text = "0"
        self.countLabel.clipsToBounds = true
        self.countLabel.layer.cornerRadius = 15
        
        self.titleLabel.frame = CGRect(x: 60, y: 0, width: screenWidth() - 150, height: 60)
        self.titleLabel.textColor = UIColor.black
        self.titleLabel.textAlignment = .center
        self.titleLabel.font = UIFont.boldSystemFont(ofSize: 17)
        self.titleLabel.text = "Please select picture"
        self.titleLabel.clipsToBounds = true
        
        self.doneButton.titleLabel?.font = UIFont.boldSystemFont(ofSize: 17)
        self.doneButton.titleLabel?.textAlignment = .right
        self.doneButton.frame = CGRect(x: screenWidth() - 75, y: 0, width: 60, height: 60)
        self.doneButton.setTitle("Done", for: .normal)
        
        self.titleViewEffectView.frame = CGRect(x: 0, y: 0, width: screenWidth(), height: 60)
        let blur = UIBlurEffect(style: .extraLight)
        self.titleViewEffectView.effect = blur
    }
    
    public var collectionView : UICollectionView!
    var imagesResource = [UIImage]()
    
    let panGesture = UIPanGestureRecognizer()
    let collectionViewPanGesture = UIPanGestureRecognizer()
    let titleViewPanGesture = UIPanGestureRecognizer()
    
    let cancelTapGesture = UITapGestureRecognizer()
    let titleViewTapGesture = UITapGestureRecognizer()
    
    var assetsFetchResults: PHFetchResult<PHAsset>!
    var imageManager: PHCachingImageManager!
    
    var selectedImagesIndex = Set<Int>()
    
    
    public override init(nibName nibNameOrNil: String?, bundle nibBundleOrNil: Bundle?) {
        super.init(nibName: nibNameOrNil, bundle: nibBundleOrNil)
        self.configure()
    }
    
    public required init?(coder: NSCoder) {
        super.init(coder: coder)
        self.configure()
    }
    
    init() {
        super.init(nibName: nil, bundle: nil)
        self.configure()
    }
    
    override public func viewDidLoad() {
        super.viewDidLoad()
        
        NotificationCenter.default.addObserver(self, selector: #selector(self.screenHasRotation), name: .UIDeviceOrientationDidChange, object: nil)
        
        self.view.addSubview(self.backView)
        self.view.addSubview(self.mainView)
        
        self.selectedImagesIndex = self.initialSelectedIndex
        self.titleView.frame = CGRect(x: 0, y: 0, width: screenWidth(), height: self.isSimpleMode ? 15 : 60)
        
        self.PangestureInit()
        self.OtherViewInit()
        self.TitleViewInit()
        // Do any additional setup after loading the view.
    }
    
    public override func viewDidAppear(_ animated: Bool) {
        //self.view.addSubview(self.mainView)

        self.isDarkMode = self.isSystemDarkMode(traitCollection: self.traitCollection)
        self.UpdateDarkMode()
        self.MainViewMoveToCenterOp(velocity: 0, isAnimated: self.isAnimated)
    }
    
    public override func traitCollectionDidChange(_ previousTraitCollection: UITraitCollection?) {
        super.traitCollectionDidChange(previousTraitCollection)
        if #available(iOS 13.0, *) {
            if traitCollection.hasDifferentColorAppearance(comparedTo: previousTraitCollection)
            {
                if self.isSwitchDarkAutomately
                {
                    self.isDarkMode = self.isSystemDarkMode(traitCollection: self.traitCollection)
                    self.UpdateDarkMode()
                }
            }
        } else {
            // Fallback on earlier versions
        }
    }
    
    var currentRotation = { () -> UIDeviceOrientation in
        switch UIDevice.current.orientation
        {
        case .landscapeLeft:
            return .landscapeLeft
        case .landscapeRight:
            return .landscapeRight
        default:
            return .portrait
        }
        
    }()
    @objc func screenHasRotation()
    {
        if !self.isSupportLandscape
        {
            return
        }
        let ori = UIDevice.current.orientation
        if ori != .landscapeLeft && ori != .landscapeRight && ori != .portrait
        {
            return
        }
        if ori != currentRotation
        {
            self.currentRotation = UIDevice.current.orientation
            self.resetViewsFrame()
            self.titleView.frame = CGRect(x: 0, y: 0, width: screenWidth(), height: self.isSimpleMode ? 15 : 60)
            
            self.UpdateCollectionView()
        }
    }
    
    let BackgroundAlpha: CGFloat = 0.5
    
    func configure()
    {
        self.modalPresentationStyle = .custom
        self.modalTransitionStyle = .coverVertical
        self.transitioningDelegate = self.transition
        self.resetAllView()
    }
    
    
    
    // MARK: - MainViewMoveOp
    
    var currentMainViewY = screenHeight()
    func MainViewMoveToCenterOp(velocity: CGFloat, isAnimated: Bool)
    {
        let distination = screenHeight() / 2.0
        let v = velocity / abs(distination - self.currentMainViewY) //Conversion unit
        self.currentMainViewY = distination
        
        if isAnimated
        {
            UIView.animate(withDuration: 0.4, delay: 0, usingSpringWithDamping: 1, initialSpringVelocity: v, options: [.allowUserInteraction, .curveEaseInOut], animations: {
                self.mainView.frame = CGRect(x: 0, y: screenHeight() / 2.0, width: screenWidth(), height: screenHeight() + 50.0)
                self.backView.alpha = self.BackgroundAlpha
            }, completion: nil)
        }
        else
        {
            self.mainView.frame = CGRect(x: 0, y: screenHeight() / 2.0, width: screenWidth(), height: screenHeight() + 50.0)
            self.backView.alpha = self.BackgroundAlpha
        }
    }
    
    func MainViewMoveToTopOp(velocity: CGFloat)
    {
        let distination = statusBarHeight
        let v = velocity / abs(distination - self.currentMainViewY) //Conversion unit
        self.currentMainViewY = distination
        self.collectionView.isScrollEnabled = true
        UIView.animate(withDuration: 0.4, delay: 0, usingSpringWithDamping: 1, initialSpringVelocity: v, options: [.allowUserInteraction, .curveEaseInOut], animations: {
            self.mainView.frame = CGRect(x: 0, y: statusBarHeight, width: screenWidth(), height: screenHeight() + 50.0)
            self.backView.alpha = self.BackgroundAlpha
        }, completion: nil)
    }
    
    func MainViewCancelOp(velocity: CGFloat)
    {
        if self.isDesideToCancel
        {
            return
        }
        self.isDesideToCancel = true
        let distination = screenHeight()
        let v = velocity / abs(distination - self.currentMainViewY) //Conversion unit
        self.currentMainViewY = distination
        UIView.animate(withDuration: 0.4, delay: 0, usingSpringWithDamping: 1, initialSpringVelocity: v, options: [.allowUserInteraction, .curveEaseInOut], animations: {
            self.mainView.frame = CGRect(x: 0, y: screenHeight(), width: screenWidth(), height: screenHeight() + 50.0)
            self.backView.alpha = 0
        }) { (true) in
            var selectedImages = [Int : UIImage]()
            for oneIndex in self.selectedImagesIndex
            {
                selectedImages[oneIndex] = self.GetImageFromIndex(item: oneIndex)
            }
            self.delegate?.imagePickerDidCancel(selectedImages)
            self.dismiss(animated: true, completion: nil)
        }
    }
    
    func MainViewMoveOp()
    {
        self.mainView.frame = CGRect(x: 0, y: self.currentMainViewY, width: screenWidth(), height: screenHeight() + 50.0)
        self.backView.alpha = self.CalculateForBackgroundAlpha(Y: self.currentMainViewY)
//        UIView.animate(withDuration: 0.1, delay: 0, options: [.allowUserInteraction, .curveEaseInOut], animations: {
//            self.mainView.frame = CGRect(x: 0, y: self.currentMainViewY, width: screenWidth(), height: screenHeight() + 50.0)
//            self.backView.backgroundColor = UIColor.init(white: 0, alpha: self.CalculateForBackgroundAlpha(Y: self.currentMainViewY))
//        }, completion: nil)
    }
    
    func CalculateForBackgroundAlpha(Y: CGFloat) -> CGFloat
    {
        if Y >= screenHeight() / 2.0 && Y <= screenHeight()
        {
            return (1.0 - (Y - screenHeight() / 2.0) / (screenHeight() / 2.0)) * self.BackgroundAlpha
        }
        else if Y > screenHeight()
        {
            return 0
        }
        else
        {
            return self.BackgroundAlpha
        }
    }
    
    func MainViewMoveToWhere(velocityY: CGFloat)
    {
        let vFast: CGFloat = 200
        
        switch self.currentMainViewY
        {
        case ..<((screenHeight() / 2.0 - statusBarHeight) / 2.0):
            if velocityY >= vFast
            {
                self.MainViewMoveToCenterOp(velocity: abs(velocityY), isAnimated: true)
            }
            else
            {
                self.MainViewMoveToTopOp(velocity: abs(velocityY))
            }
        case ((screenHeight() / 2.0 - statusBarHeight) / 2.0) ... screenHeight() / 2.0:
            if velocityY <= -vFast
            {
                self.MainViewMoveToTopOp(velocity: abs(velocityY))
            }
            else
            {
                self.MainViewMoveToCenterOp(velocity: abs(velocityY), isAnimated: true)
            }
        case screenHeight() / 2.0 ... screenHeight() * 3.0 / 4.0:
            if velocityY >= vFast
            {
                self.MainViewCancelOp(velocity: abs(velocityY))
            }
            else
            {
                self.MainViewMoveToCenterOp(velocity: abs(velocityY), isAnimated: true)
            }
        default:
            if velocityY <= -vFast
            {
                self.MainViewMoveToCenterOp(velocity: abs(velocityY), isAnimated: true)
            }
            else
            {
                self.MainViewCancelOp(velocity: abs(velocityY))
            }
        }
    }
    
    
    
    // MARK: - TitlePanGesture
    
    func TitlePanGestureInit()
    {
        self.titleViewPanGesture.delegate = self
        self.titleViewPanGesture.addTarget(self, action: #selector(titlePanned(_:)))
        self.titleView.addGestureRecognizer(self.titleViewPanGesture)
    }
    
    var titleViewStartY: CGFloat = 0
    @objc func titlePanned(_ gesture : UIPanGestureRecognizer)
    {
        switch gesture.state
        {
        case .began:
            self.titleViewStartY = self.currentMainViewY
        case .changed:
            let gestureY = self.titleViewStartY + gesture.translation(in: gesture.view).y
            self.currentMainViewY = gestureY <= statusBarHeight ? statusBarHeight : gestureY
            self.MainViewMoveOp()
        case .ended:
            let velocityY = gesture.velocity(in: gesture.view).y
            self.MainViewMoveToWhere(velocityY: velocityY)
        case .cancelled, .failed:
            self.MainViewMoveToWhere(velocityY: 0)
        default:
            break
        }
    }
    
    // MARK: - PanGesture
    
    func PangestureInit()
    {
        self.panGesture.delegate = self
        self.panGesture.addTarget(self, action: #selector(Panned(_:)))
        self.mainView.addGestureRecognizer(self.panGesture)
        
        self.cancelTapGesture.addTarget(self, action: #selector(viewTapped(_:)))
        self.backView.addGestureRecognizer(cancelTapGesture)
    }
    
    @objc func viewTapped(_ gesture : UITapGestureRecognizer)
    {
        self.MainViewCancelOp(velocity: 0)
    }
    
    var MainViewStartY: CGFloat?
    @objc func Panned(_ gesture : UIPanGestureRecognizer)
    {
        if self.collectionView.isScrollEnabled
        {
            return
        }
        switch gesture.state
        {
        case .began:
            self.MainViewStartY = self.currentMainViewY
        case .changed:
            self.MainViewStartY = self.MainViewStartY == nil ? self.currentMainViewY - gesture.translation(in: gesture.view).y : self.MainViewStartY
            let gestureY = self.MainViewStartY! + gesture.translation(in: gesture.view).y
            self.currentMainViewY = gestureY <= statusBarHeight ? statusBarHeight : gestureY
            self.MainViewMoveOp()
        case .ended:
            let velocityY = gesture.velocity(in: gesture.view).y
            self.MainViewMoveToWhere(velocityY: velocityY)
            self.MainViewStartY = nil
        case .cancelled, .failed:
            self.MainViewMoveToWhere(velocityY: 0)
            self.MainViewStartY = nil
        default:
            break
        }
    }
    
    
    // MARK: - OtherView
    
    func OtherViewInit()
    {
        self.collectionViewInit()
    }

    /*
    // MARK: - Navigation

    // In a storyboard-based application, you will often want to do a little preparation before navigation
    override func prepare(for segue: UIStoryboardSegue, sender: Any?) {
        // Get the new view controller using segue.destination.
        // Pass the selected object to the new view controller.
    }
    */

}


